ffmpeg/lpms_ffmpeg.c

```diff
@@ -102,3 +140,588 @@ int lpms_rtmp2hls(char *listen, char *outf, char *ts_tmpl, char* seg_time)
+    pkt.duration = av_rescale_q(pkt.duration, encoder->time_base, ost->time_base);
+  }
+
+  ret = av_interleaved_write_frame(octx->oc, &pkt);
+  if (ret < 0) proc_err("Error writing frame\n"); // XXX handle better?
+
+proc_cleanup:
+  av_frame_unref(frame);
+  av_packet_unref(&pkt);
+  return ret;
+#undef proc_err
+}
+
+#define MAX_OUTPUT_SIZE 3
+
+int lpms_transcode(char *inp, output_params *params, int nb_outputs)
+{
+#define main_err(msg) { \
+  if (!ret) ret = -1; \
+  fprintf(stderr, msg); \
+  goto transcode_cleanup; \
+}
+  int ret = 0, i = 0;
+  struct input_ctx  ictx = {0};
+  AVPacket ipkt = {0};
+  char errbuf[AV_ERROR_MAX_STRING_SIZE];
+  struct output_ctx outputs[MAX_OUTPUT_SIZE];
+  AVFrame *dframe;
+  memset(outputs, 0, sizeof outputs);
+  ret = open_input(inp, &ictx);
+  if (ret < 0) main_err("transcoder: Unable to open input\n");
+  for (i = 0; i < nb_outputs; i++) {
+    struct output_ctx *octx = &outputs[i];
+    octx->fname = params[i].fname;
+    octx->width = params[i].w;
+    octx->height = params[i].h;
+    if (params[i].bitrate) octx->bitrate = params[i].bitrate;
+    if (params[i].fps.den) octx->fps = params[i].fps;
+    if (ictx.vc) {
+      char filter_str[256];
+      snprintf(filter_str, sizeof filter_str, "fps=fps=%d/%d,scale=%dx%d:force_original_aspect_ratio=decrease", octx->fps.num, octx->fps.den, octx->width, octx->height);
+      ret = init_video_filters(&ictx, octx, filter_str);
+      if (ret < 0) main_err("Unable to open video filter");
+    }
+    if (ictx.ac) {
+      char filter_str[256];
+      //snprintf(filter_str, sizeof filter_str, "aformat=sample_fmts=s16:channel_layouts=stereo:sample_rates=44100,asetnsamples=n=1152,aresample");
+      snprintf(filter_str, sizeof filter_str, "aformat=sample_fmts=fltp:channel_layouts=stereo:sample_rates=44100"); // set sample format and rate based on encoder support
+      ret = init_audio_filters(&ictx, octx, filter_str);
+      if (ret < 0) main_err("Unable to open audio filter");
+    }
+    ret = open_output(octx, &ictx);
+    if (ret < 0) main_err("transcoder: Unable to open output\n");
+  }
+  av_init_packet(&ipkt);
+  dframe = av_frame_alloc();
+  if (!dframe) main_err("transcoder: Unable to allocate frame\n");
+
+  while (1) {
+    AVStream *ist = NULL;
+    ret = process_in(&ictx, dframe, &ipkt);
+    if (ret == AVERROR_EOF) break;
+    else if (ret < 0) goto whileloop_end; // XXX fix
+    ist = ictx.ic->streams[ipkt.stream_index];
+
+    for (i = 0; i < nb_outputs; i++) {
+      struct output_ctx *octx = &outputs[i];
+      struct filter_ctx *filter = NULL;
+      AVStream *ost = NULL;
+      AVCodecContext *encoder = NULL;
+
+      if (ist->index == ictx.vi && ictx.vc) {
+        ost = octx->oc->streams[0];
+        encoder = octx->vc;
+        filter = &octx->vf;
+      } else if (ist->index == ictx.ai && ictx.ai) {
+        ost = octx->oc->streams[!!ictx.vc];
```

Comment:
Why !!ictx.vc here instead of something like `octx->oc->streams[1]`?

Reply:
Accommodates the case where we don't have video. Video streams are always at index 0, but audio may be 0 or 1 depending on whether `ictx.vc` is non-null.

Reply:
Added a comment explaining as much, and test cases to ensure that it actually handles no-audio and no-video correctly. Note that this line contains a bug:

`} else if (ist->index == ictx.ai && ictx.ai) {`

which should be `} else if (ist->index == ictx.ai && ictx.ac) {`